### PR TITLE
feat: close feature gap by wiring up per-tensor routed FP8 fused-moe

### PR DESCRIPTION
**Pull request by jdebache**
## 📌 Description

FP8 per-tensor routed fused-MoE is supported at the kernel level. This PR exposes it and tests it.

### ✅ Pre-commit Checks

- [X] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [X] I have installed the hooks with `pre-commit install`.
- [X] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## 🧪 Tests

- [X] Tests have been added or updated as needed.
- [X] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

- **New Features**
  - Added FP8 per-tensor MoE pre-routed execution that accepts packed expert top-k indices (routing logits can be omitted).
  - Introduced a new routed FP8 per-tensor MoE API and exposed it at the Python package top level.
  - Supports non-zero local expert offsets for correct shifted expert dispatch.
- **Tests**
  - Added parity tests comparing logits-based vs pre-routed outputs, including renormalize variants.
  - Added correctness coverage for non-zero expert offset scenarios.
- **Trace**
  - Added routed MoE trace template, example output, and trace-schema/reference tests.

### CODE REVIEW

File: csrc/trtllm_fused_moe_kernel_launcher.cu [hunk 1]
```diff
@@ -2308,8 +2394,101 @@ Array<Tensor> trtllm_fp8_per_tensor_scale_moe(
...
+    int64_t local_expert_offset, int64_t local_num_experts, Optional<double> routed_scaling_factor,
+    bool use_routing_scales_on_input, int64_t routing_method_type, bool do_finalize,
+    bool enable_pdl, Array<int64_t> config_index, int64_t activation_type, bool norm_topk_prob,
+    Optional<TensorView> routing_replay_out) {
+  // Basic type validation
+  auto const dtype = hidden_states.dtype();
+  auto const activation = validateAndCastActivationType(activation_type);
+
+  TVM_FFI_ICHECK_EQ(expert_indices.dtype(), dl_int32)
+      << "FP8 MoE: expert_indices must be int32 (packed expert_id/weight).";
```
Comment (feih-nv) [inline csrc/trtllm_fused_moe_kernel_launcher.cu:2435]: This comment has not been fully addressed. 

Could we validate `expert_indices` fully at the start of this routed API?

This routed entry point has no routing logits, so it must always receive packed expert indices with shape `[num_tokens, top_k]`.

Currently, inputs such as these are not rejected clearly:

- a 1-D tensor such as `[num_tokens * top_k]`
- a 3-D tensor
- an empty tensor such as `[0, top_k]` when `hidden_states` contains tokens
- a tensor with the wrong top-k width, such as `[num_tokens, top_k + 1]`
- a noncontiguous tensor
- a CPU tensor or a tensor on a different GPU

Some malformed shapes cause `has_precomputed_routing()` to return false. The launcher then behaves as if no packed routing was supplied, but this API also has no logits, so the routing kernel may read an uninitialized buffer.

Please require:

```text
- CUDA tensor on the same device as `hidden_states`
- contiguous `int32`
- shape `[num_tokens, top_k]`
```

File: flashinfer/fused_moe/core.py [hunk 2]
```diff
@@ -3571,6 +3778,141 @@ def trtllm_fp8_per_tensor_scale_moe(
         return result
 
 
+@flashinfer_api
+def trtllm_fp8_per_tensor_scale_routed_moe(
```
Comment (feih-nv) [inline flashinfer/fused_moe/core.py:3784]: Could this new public API be wired to a trace template?

Repository guidance requires public `@flashinfer_api` entry points to provide `trace=...`, add an example in `tests/trace/example.py`, regenerate the trace JSON, and run `tests/trace/`.

The routed per-tensor API currently has no trace template/example, so `fi_trace` cannot capture this path.

File: tests/moe/test_trtllm_gen_routed_fused_moe.py [hunk 3]
```diff
@@ -458,6 +460,134 @@ def test_trtllm_gen_fp8_routed_fused_moe(
...
+        output1_scales_scalar=output1_scales_scalar,
+        output1_scales_gate_scalar=output1_scales_gate_scalar,
+        gemm2_weights=gemm2_weights,
+        output2_scales_scalar=output2_scales_scalar,
+        num_experts=num_experts,
+        top_k=top_k,
+        n_group=None,
+        topk_group=None,
+        intermediate_size=intermediate_size,
+        local_expert_offset=0,
```
Comment (feih-nv) [inline tests/moe/test_trtllm_gen_routed_fused_moe.py:576]: Could we add a sharded expert-offset case perhaps?

File: flashinfer/fused_moe/core.py [hunk 4]
```diff
@@ -2219,6 +2248,183 @@ def _fake_trtllm_fp8_per_tensor_scale_moe(
...
+            enable_pdl,
+            [-1, -1] if tactic == -1 else tactic,
+            activation_type,
+            True,  # norm_topk_prob: unused for pre-computed routing
+            routing_replay_out,
+        )
+        return _unpack_trtllm_moe_output(intermediate_output, output, do_finalize, None)
+
+    @register_fake_op("flashinfer::trtllm_fp8_per_tensor_scale_routed_moe")
+    def _fake_trtllm_fp8_per_tensor_scale_routed_moe(
```
Comment (feih-nv) [inline flashinfer/fused_moe/core.py:2393]: How should the fake op handle `do_finalize=False`?

The real routed op returns three tensors when finalization is disabled:
```text
gemm2_output
expert_weights
expanded_idx_to_permuted_idx
```
but the fake op always returns one `[num_tokens, hidden_size]` tensor.

This is also tracked by #3926 , here we may do 2 at least?

1. make its fake output match the real three-tensor result, or
2. explicitly reject do_finalize=False?

**Review by feih-nv — APPROVED** [review]

**Comment by feih-nv** [comment]
Retriggered the failed 5090 job in the pipeline above, passed here https://gitlab-master.nvidia.com/dl/flashinfer/flashinfer-ci/-/jobs/374648320.
